src/Storages/MergeTree/MergeTreeReadPoolBase.cpp

```diff
@@ -113,9 +118,25 @@ MergeTreeReadTaskPtr MergeTreeReadPoolBase::createTask(
         ? std::make_unique<MergeTreeBlockSizePredictor>(*read_info->shared_size_predictor)
         : nullptr; /// make a copy
 
-    auto get_part_name = [](const auto & task_info) -> const String &
+    auto get_part_name = [](const auto & task_info) -> String
     {
-        return task_info.data_part->isProjectionPart() ? task_info.data_part->getParentPart()->name : task_info.data_part->name;
+        const auto & data_part = task_info.data_part;
+
+        if (data_part->isProjectionPart())
```

Comment:
@kssenii what do you think about this as a help for catching when we actually lose the parent part?

Even though we don't use the part itself, it can be useful getting the name. If you think it's too much I can revert to just using `getParentPartName`.

cc @tavplubix

Reply:
Looks good to me.

Reply:
Btw as error happens not only with projections, I had a thought may be we should pass part's shared pointers from `ReadFromMergeTree`, e.g. `prepared_parts` value and save in `MergeTreePrefetchedReadPool` when we create it in `readFromPool`, may be it will help.

Reply:
@nickitat you added a slightly different check in private, please verify if I can replace it with this